flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaEvolutionUtil.java

```diff
@@ -62,4 +88,240 @@ public static int[] createIndexMapping(
+     * follows
+     *
+     * <ul>
+     *   <li>table fields: 1->c, 3->a, 4->e, 5->d, 6->b
+     *   <li>data fields: 1->a, 2->b, 3->c, 4->d
+     * </ul>
+     *
+     * <p>When we project 1->c, 6->b, 3->a from table fields, the table projection is [[0], [4],
+     * [1]], in which 0 is the index of field 1->c, 4 is the index of field 6->b, 1 is the index of
+     * field 3->a in table fields. We need to create data projection from [[0], [4], [1]] as
+     * follows:
+     *
+     * <ul>
+     *   <li>Get field id of each index in table projection from table fields
+     *   <li>Get index of each field above from data fields
+     * </ul>
+     *
+     * <p>The we can create table projection as follows: [[0], [-1], [2]], in which 0, -1 and 2 are
+     * the index of fields [1->c, 6->b, 3->a] in data fields. When we project column from underlying
+     * data, we need to specify the field index and name. It is difficult to assign a proper field
+     * id and name for 6->b in data projection and add it to data fields, and we can't use 6->b
+     * directly because the field index of b in underlying is 2. We can remove the -1 field index in
+     * data projection, then the result data projection is: [[0], [2]].
+     *
+     * <p>We create {@link RowData} for 1->a, 3->c after projecting them from underlying data, then
+     * create {@link ProjectedRowData} with a index mapping and return null for 6->b in table
+     * fields.
+     *
+     * @param tableFields the fields of table
+     * @param dataFields the fields of underlying data
+     * @param tableProjection the projection of table
+     * @return the projection of data
+     */
+    public static int[][] createDataProjection(
+            List<DataField> tableFields, List<DataField> dataFields, int[][] tableProjection) {
+        List<Integer> dataFieldIdList =
+                dataFields.stream().map(DataField::id).collect(Collectors.toList());
+        return Arrays.stream(tableProjection)
+                .map(p -> Arrays.copyOf(p, p.length))
+                .peek(
+                        p -> {
+                            int fieldId = tableFields.get(p[0]).id();
+                            p[0] = dataFieldIdList.indexOf(fieldId);
+                        })
+                .filter(p -> p[0] >= 0)
+                .toArray(int[][]::new);
+    }
+
+    /**
+     * Create predicate list from data fields. We will visit all predicate in filters, reset it's
+     * field index, name and type, and use {@link AlwaysFalse} or {@link AlwaysTrue} if the field is
+     * not exist.
+     *
+     * @param tableFields the table fields
+     * @param dataFields the underlying data fields
+     * @param filters the filters
+     * @return the data filters
+     */
+    @Nullable
+    public static List<Predicate> createDataFilters(
+            List<DataField> tableFields, List<DataField> dataFields, List<Predicate> filters) {
```

Comment:
Maybe we can construct them first like:
- `Map<String, DataField> tableFields`?
- `LinkedHashMap<Integer, DataField> dataFields` and we can create a `indexOf` for `LinkedHashMap`?